### PR TITLE
prov/verbs: fixed init issue

### PR DESCRIPTION
**Pull request by hoopoepg**
in some configurations verbs provider missing initialization
of one of domain (RDM domain could not attach to iface)

update iface detection and binding

Signed-off-by: Oblomov, Sergey <sergey.oblomov@intel.com>

### CODE REVIEW

File: prov/verbs/src/verbs_info.c [hunk 1]
```diff
@@ -849,21 +849,15 @@ static int fi_ibv_copy_ifaddr(const char *name, const char *service, uint64_t fl
...
-		for (fi = info; fi; fi = fi->next)
+		for (fi = info; fi; fi = fi->next) {
 			if (!strncmp(id->verbs->device->name, fi->domain_attr->name,
-						strlen(id->verbs->device->name)))
-				break;
-		if (!fi) {
-			FI_WARN(&fi_ibv_prov, FI_LOG_FABRIC,
-					"No matching fi_info for device: "
-					"%s with address: %s\n",
-					id->verbs->device->name, name);
```
Comment (hoopoepg) [inline prov/verbs/src/verbs_info.c]: hi
for now it processes all valid interfaces, and ignores all other

Comment (a-ilango) [inline prov/verbs/src/verbs_info.c]: @hoopoepg , I'm not sure if I understand you. The warning is raised when we have a valid verbs interface corresponding to id->verbs but no matching fi_info based on the domain attribute name. Are you seeing this warning for an interface that's not a valid verbs device?

Comment (hoopoepg) [inline prov/verbs/src/verbs_info.c]: I see your point. Will update PR tomorrow

File: prov/verbs/src/verbs_info.c [hunk 2]
```diff
@@ -502,8 +506,29 @@ int fi_ibv_fi_to_rai(const struct fi_info *fi, uint64_t flags,
...
+
+		rai->ai_family = addr->sin_family;
+		rai->ai_flags |= RAI_FAMILY;
+
+		rai->ai_src_len = sizeof(*addr);
+		if (!(rai->ai_src_addr = malloc(rai->ai_src_len)))
+			return -FI_ENOMEM;
+		memcpy(rai->ai_src_addr, addr, rai->ai_src_len);
 		return 0;
+	}
```
Comment (hoopoepg) [inline prov/verbs/src/verbs_info.c]: it is black magic in verbs initialization :)

I added fi_dup on every interface detected on system.

**Comment by a-ilango** [comment]
Was there any issue with adding a duplicate fi_info just when you find the source address set (to account for extra interface)?

**Comment by hoopoepg** [comment]
nop, but on some systems there are duplicated interfaces & it creates duplicates of info

**Comment by a-ilango** [comment]
> nop, but on some systems there are duplicated interfaces & it creates duplicates of info

If by duplicate interface you mean the virtual interface, isn't that what we want? An extra fi_info which corresponds to the virtual interface with src_addr set correctly would help in not overwriting the src_addr of the real interface.

**Comment by hoopoepg** [comment]
Every interface has more than one address, for example ip4 and ip6. Such addresses are reflected in separate copies of fi_info. And for cases when user didn't set interface name it creates info for all interfaces

**Comment by shefty** [comment]
Does this PR need to be updated to account for #2641 now?

**Comment by a-ilango** [comment]
I believe this PR can be closed since changes in #2641 handles additional interfaces now. @hoopoepg can confirm. The interface filtering logic might need to be updated so that it doesn't remove any fi_info unnecessarily. That can be done in a new PR.

@jozie would you be able to verify if #2556 is resolved on libfabric master?

**Comment by hoopoepg** [comment]
hi
yes, I tested this patch on system where 2 ib interfaces configured - it resolves all issues
thank you

**Comment by JoZie** [comment]
Hi

Yes my issue #2556 is fixed now. Thank you for that!

But I find it odd that with 2 ib interfaces `fi_info` shows only 3 verbs providers, 2 `FI_EP_MSG` and one of type `FI_EP_RDM`. I would assume that there should be 4 providers displayed.  I wanted to test if this changes if I provide an IP with the `-n` argument but this causes `fi_info` to crash with a `SIGSEGV` -- the passed `hints` pointer is `NULL`.  (new bug?)

```
fi_ibv_getinfo (version=65540, node=0x7ffffd74f81d "172.24.16.16", service=0x0, flags=0, hints=0x0, info=0x7ffffd74daa0) at ../prov/verbs/src/verbs_info.c:1364
```

After this I tried the following commands but they returned an error code or so. So there are still issues.
```
fi_info -t FI_EP_RDM -n 172.24.0.16
fi_getinfo: -61

fi_info -t FI_EP_RDM -n 172.24.16.16
fi_getinfo: -61
```

**Comment by a-ilango** [comment]
@JoZie  thanks for testing the fix.

>But I find it odd that with 2 ib interfaces fi_info shows only 3 verbs providers, 2 FI_EP_MSG and one of type FI_EP_RDM. I would assume that there should be 4 providers displayed.

Yes, this is because the filtering logic in fi_ibv_rdm_remove_nonaddr_info removes the fi_info corresponding to one of the interfaces. @hoopoepg do you have plans to update the filtering logic to fix this issue?

>if I provide an IP with the -n argument but this causes fi_info to crash with a SIGSEGV -- the passed hints pointer is NULL. (new bug?)

Yes, this is a bug. I'll prepare a PR to fix it.

>fi_info -t FI_EP_RDM -n 172.24.0.16
>fi_getinfo: -61

I'll fix this as well.

**Comment by hoopoepg** [comment]
> @hoopoepg do you have plans to update the filtering logic to fix this issue?
hi
sure, I will look what is wrong (on next week, ok?)

**Comment by a-ilango** [comment]
Closing this PR as the original issue has been fixed by #2641

**Pull request closed without merging**